### PR TITLE
Add support for extends to typescript config and babel config parsers

### LINKED ISSUE #943: In babel and typescript parsers, build the full configuration when they are using extends.
Right now, we just read and parse the configuration files, but we need to build the final configuration before check if it pass the rules

### PR DESCRIPTION
**Pull request by sarvaje**
Fix #943

## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/sonarwhal/sonarwhal)
- [x] Followed the [commit message guidelines](https://sonarwhal.com/docs/contributor-guide/getting-started/pull-requests/#commit-messagess)

For non-trivial changes, please make sure you also:

- [x] Added/Updated related documentation.
- [x] Added/Updated related tests.

### CODE REVIEW

File: packages/parser-typescript-config/src/parser.ts [hunk 1]
```diff
@@ -67,6 +71,28 @@ export default class TypeScriptConfigParser extends Parser {
         try {
             config = JSON.parse(fetchEnd.response.body.content);
 
+            const originalConfig = cloneDeep(config);
+
+            try {
```
Comment (molant) [inline packages/parser-typescript-config/src/parser.ts]: Nested `try`? Isn't there a way to refactor this?

File: packages/parser-typescript-config/src/parser.ts [hunk 2]
```diff
@@ -67,6 +71,28 @@ export default class TypeScriptConfigParser extends Parser {
...
+                config = finalConfig(config, resource);
+            } catch (err) {
+                if (err.code === ErrorCodes.circular) {
```
Comment (molant) [inline packages/parser-typescript-config/src/parser.ts]: This checks is also in Babel. Can't it be refactored and maybe added to the `Parser` class so it's available in all `Parsers`?

**Comment by sarvaje** [comment]
@molant I have moved some code to `Parser` and I have added some tests and modify the docs

**Comment by molant** [comment]
@sarvaje looks like there are some errors with the tests:

```
rule-typescript-config › dist › tests › is-valid › [local] If the configuration has a circular reference, it should fail Different message
parser-javascript › dist › tests › tests › If an script tag is an internal javascript, then we should parse the code and emit a parse::javascript::end event Rejected promise returned by test
parser-javascript › dist › tests › tests › If fetch::end::script is received, then we should parse the code and emit a parse::javascript::end event Rejected promise returned by test
```

File: packages/parser-babel-config/README.md [hunk 3]
```diff
@@ -43,7 +43,19 @@ This `parser` emits the following events:
...
+  * `resource`: the parsed resource.
+  * `error`: the error emited parsing the configuration file.
+
+* `parse::babel-config::error::circular`, of type `BabelConfigInvalidJSON`
+  which contains the following information:
+
+  * `resource`: the parsed resource.
+  * `error`: the error emited parsing the configuration file.
+
+* `parse::babel-config::error::extends`, of type `BabelConfigInvalidJSON`
```
Comment (alrra) [inline packages/parser-babel-config/README.md]: Note: In a conversation with @molant we talked about not having the `...error::not-found` (or similar events) as most of the time just one rule will needed it, and in that case, the rule can have that logic in it.

File: packages/sonarwhal/docs/contributor-guide/how-to/parser.md [hunk 4]
```diff
@@ -13,7 +13,7 @@ import { Sonarwhal } from '../../sonarwhal';
 export default class CustomParser extends Parser {
 
     public constructor(sonarwhal: Sonarwhal) {
-        super(sonarwhal);
+        super(sonarwhal, 'parser-name');
```
Comment (alrra) [inline packages/sonarwhal/docs/contributor-guide/how-to/parser.md]: For consistency, maybe use:

> `<parser_name>`

File: packages/sonarwhal/docs/contributor-guide/how-to/parser.md [hunk 5]
```diff
@@ -26,7 +26,7 @@ export default class CustomParser extends Parser {
         // and maybe leter use a schema if a configuration file or something else
 
         // If there's something to share, do it via an event
-        await this.sonarwhal.emitAsync('customparser::custom', data);
+        await this.sonarwhal.emitAsync('parse::parser-name::<end | error>', data);
```
Comment (alrra) [inline packages/sonarwhal/docs/contributor-guide/how-to/parser.md]: > `<parser_name>` ?

File: packages/sonarwhal/docs/contributor-guide/how-to/parser.md [hunk 6]
```diff
@@ -40,7 +40,7 @@ Once you have analyzed the resource, the way to share information is via
 events (custom or not):
 
 ```ts
-await this.sonarwhal.emitAsync('customparser::custom', data);
+await this.sonarwhal.emitAsync('parse::parse-name::<end | error>', data);
```
Comment (alrra) [inline packages/sonarwhal/docs/contributor-guide/how-to/parser.md]: > `<parser_name>` ?

File: packages/parser-babel-config/README.md (hunk 3 above)
Comment (sarvaje) [reply packages/parser-babel-config/README.md]: Can you open an issue for that?

**Comment by sarvaje** [comment]
@alrra done!

Also, tests should be fixed.

File: packages/parser-babel-config/README.md (hunk 3 above)
Comment (molant) [reply packages/parser-babel-config/README.md]: https://github.com/sonarwhal/sonarwhal/issues/960

**Pull request merged**